enable mkldnn conv2d backward to support mkldnn tensor input

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37218 BFloat16: enable prepacked weights's inference
* #37215 BFloat16: add explicit dtype support for to_mkldnn and to_dense
* #37199 add mkldnn relu backward and reshape backward
* #37147 add mkldnn batch_norm backward
* #37146 add mkldnn pooling backward
* #36122 add mkldnn linear backward
* **#36121 enable mkldnn conv2d backward to support mkldnn tensor input**

Differential Revision: [D22440969](https://our.internmc.facebook.com/intern/diff/D22440969)

XiaobingSuper committed:
enable mkldnn conv2d backward to support mkldnn tensor input

[ghstack-poisoned]

XiaobingSuper commented:
@VitalyFedyunin , those PRs https://github.com/pytorch/pytorch/pull/20567 https://github.com/pytorch/pytorch/pull/20570 https://github.com/pytorch/pytorch/pull/20571 https://github.com/pytorch/pytorch/pull/20572 are tool old , I will re-pull them. thanks!

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

XiaobingSuper commented:
@uyongw, @Jianhui-Li, @jgong5, @hongzhen1.

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

XiaobingSuper commented:
@VitalyFedyunin , could you help review this code?

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

VitalyFedyunin commented:
Are we fine with BC-incompatible changes in `mkldnn_convolution_backward_weights` operator?

What is the purpose of ideep update?

XiaobingSuper commented:
The purpose of ideep update:

1. support conv3d forward and backward
2.  solve group convolution backward perfromance degradation
3. solve batchnorm backward issue when input is nchw4c

Why I change **mkldnn_convolution_backward_weights**?  for bf16 backward path, the weight can be a fp32(just convert input to bfloat16 MKLDNN tensor) or bf16 tensor(convert input to a bf16 MKLDNN tensor and call model.to_mkldnn(torch.bfloat16)), so I need get the data_type from this weight parameters.

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

VitalyFedyunin commented:
Test errors seems to be mkldnn related
```
Traceback (most recent call last):
  File "test_mkldnn.py", line 149, in test_conv2d
    rtol=1e-5)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 928, in assertEqual
    assertTensorsEqual(x, y)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 890, in assertTensorsEqual
    torch.testing.assert_allclose(a, b, atol=atol, rtol=rtol, equal_nan=True, msg=message)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\__init__.py", line 60, in assert_allclose
    raise AssertionError(msg)
AssertionError: Not within tolerance rtol=1e-05 atol=1e-05 at input[0, 1, 1, 2] (-1572.8182373046875 vs. -1572.835693359375) and 0 other locations (5.00%)
```

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

VitalyFedyunin commented:
Please whitelist changed operator in https://github.com/pytorch/pytorch/blob/70192c651ce247f01de594d5ceeb74bfdee27cfb/test/backward_compatibility/check_backward_compatibility.py#L19

VitalyFedyunin commented on test/test_mkldnn.py:
```diff
@@ -110,22 +110,48 @@ def test_conv2d(self):
-                mkldnn_conv2d = mkldnn_utils.to_mkldnn(copy.deepcopy(conv2d))
+            for train in [True, False]:
+                for bias in [True, False]:
+                    conv2d = torch.nn.Conv2d(in_channels=C,
```
This looks like indent problem for me.

VitalyFedyunin commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1973,7 +1973,7 @@
   use_c10_dispatcher: full
 
-- func: mkldnn_convolution_backward_weights(int[] weight_size, Tensor grad_output, Tensor self, int[] padding, int[] stride, int[] dilation, int groups, bool bias_defined) -> (Tensor, Tensor)
+- func: mkldnn_convolution_backward_weights(Tensor weight, Tensor grad_output, Tensor self, int[] padding, int[] stride, int[] dilation, int groups, bool bias_defined) -> (Tensor, Tensor)
```
Please put inline comment here, why we need weight to be tensor in compare to cudnn and miopen declarations.

XiaobingSuper replied:
done.

VitalyFedyunin reviewed (changes requested):
Overall looks good, need some tests alterations.

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"




[ghstack-poisoned]

VitalyFedyunin commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -115,6 +115,7 @@
     ('aten::__or__', datetime.date(2020, 6, 30)),
     ('aten::__xor__', datetime.date(2020, 6, 30)),
     ('aten::split', datetime.date(2020, 6, 30)),
+    ('aten::mkldnn_convolution_backward_weights', datetime.date(2020, 6, 30)),
```
Sorry you need to update this date before landing.

XiaobingSuper replied:
changed.

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D22440969](https://our.internmc.facebook.com/intern/diff/D22440969)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D22440969](https://our.internmc.facebook.com/intern/diff/D22440969)

[ghstack-poisoned]

VitalyFedyunin commented:
Land fails with:

```
Command failed with exit code 1.

stderr: caffe2/aten/src/ATen/native/mkldnn/Conv.cpp:159:5: error: no matching function for call to 'compute'
    ideep::convolution_backward_weights::compute(
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 [removed]/build/ideep/include/ideep/operators/conv.hpp:476:15: note: candidate function not viable: no known conversion from 'dnnl::memory::data_type' to 'ideep::algorithm' (aka 'dnnl::algorithm') for 11th argument
  static void compute(const tensor& src,
              ^
[removed]/build/ideep/include/ideep/operators/conv.hpp:493:15: note: candidate function not viable: no known conversion from 'ideep::tensor' to 'const ideep::dims' (aka 'const vector<long>') for 5th argument
  static void compute(const tensor& src,
              ^
caffe2/aten/src/ATen/native/mkldnn/Conv.cpp:172:5: error: no matching function for call to 'compute'
    ideep::convolution_backward_weights::compute(
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[removed]/build/ideep/include/ideep/operators/conv.hpp:493:15: note: candidate function not viable: no known conversion from 'dnnl::memory::data_type' to 'ideep::algorithm' (aka 'dnnl::algorithm') for 10th argument
  static void compute(const tensor& src,
              ^
[removed]/build/ideep/include/ideep/operators/conv.hpp:476:15: note: candidate function not viable: no known conversion from 'std::vector<long>' to 'ideep::tensor &' for 5th argument
  static void compute(const tensor& src,
              ^
2 errors generated.
```

XiaobingSuper commented:
@VitalyFedyunin ,  this error seems not find the corresponding function, I can reproduce this error by checkout old ideep  commit[(ca7b718)](https://github.com/intel/ideep/commit/ca7b718d550e869e09c2a63d258b4aebe2aed91f), could you help check the ideep commit id which used by the failed case?  PyTorch master use [938cc68](https://github.com/intel/ideep/commit/938cc68897bb46b8d4b228966edd9e23e471cf3b) now. Thanks!

ZhuJewel commented:
@VitalyFedyunin , please let us know if you need any help for the errors related to the ideep. Thanks.

VitalyFedyunin commented:
Hi! I'm ready to land it. Can you please rebase to make sure we avoid merge conflicts.

jgong5 commented:
> Hi! I'm ready to land it. Can you please rebase to make sure we avoid merge conflicts.

@VitalyFedyunin Thank you. We created a new PR https://github.com/pytorch/pytorch/pull/48994 to ease the rebase. I copied you there. This old one is supposed to be closed. Thanks.

VitalyFedyunin commented:
Are you planning to port the remaining 4 PRs?

Jianhui-Li commented:
> Are you planning to port the remaining 4 PRs?

I think so. @XiaobingSuper
